app/models/assignment.rb

```diff
@@ -600,6 +600,59 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generates the summary of the most test results associated with an assignment.
+  def summary_test_results
+    test_groups_query = self.test_groups
+                            .joins(:test_group_results)
+                            .group(:id)
+                            .select('test_groups.id AS test_groups_id',
+                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
+                            .to_sql
+    test_group_query = TestGroupResult.joins(:test_group)
+                                      .joins("INNER JOIN (#{test_groups_query}) sub \
+                                              ON test_groups.id = sub.test_groups_id \
+                                              AND test_group_results.created_at = sub.test_group_results_created_at")
+                                      .select('test_groups.name AS test_group_name', 'test_group_results.*')
+                                      .to_sql
+
+    TestResult.joins("INNER JOIN (#{test_group_query}) sub ON test_results.id = sub.test_run_id")
+              .select('test_results.name as test_result_name',
+                      'test_results.output',
+                      'test_results.marks_earned as test_result_marks_earned',
+                      'test_results.marks_total as test_result_marks_total',
+                      'test_results.id as id',
+                      :status, :marks_earned, :marks_total, :extra_info, :error_type)
+  end
+
+  # Generate a JSON summary of the most recent test results associated with an assignment.
+  def summary_test_result_json
+    self.summary_test_results.to_json
```

Comment:
Okay, so right now the JSON structure is a flat list, but looking at this more carefully, I'd like to see some grouping of the results, first by grouping id, and then by test group name. So the overall JSON should look like:

```
{
  <grouping_id>:
    {
      <test group name>: [<test result 1>, <test result 2>, ...],
      <test group name>: [<test result 1>, <test result 2>, ...],
      ...
    },
  ...
}
```
Note that you can use Ruby's [`Enumerable#group_by`](https://ruby-doc.org/core-3.0.2/Enumerable.html#method-i-group_by) to do a lot of the grouping for you, just be careful when handling the nested structure.